pkg/networkserver/adr.go

```diff
@@ -157,13 +199,11 @@ func adaptDataRate(dev *ttnpb.EndDevice, fps *frequencyplans.Store, defaults ttn
 	}
 
 	if len(ups) >= 2 {
-		lossRate := float32(up.Payload.GetMACPayload().FHDR.FCnt-ups[0].Payload.GetMACPayload().FHDR.FCnt-uint32(len(ups))) /
-			float32(up.Payload.GetMACPayload().FHDR.FCnt-ups[0].Payload.GetMACPayload().FHDR.FCnt)
-		switch {
-		case lossRate < 0.05:
+		switch r := lossRate(dev.MACState.CurrentParameters.ADRNbTrans, ups...); {
+		case r < 0.05:
 			dev.MACState.DesiredParameters.ADRNbTrans = 1 + dev.MACState.DesiredParameters.ADRNbTrans/3
-		case lossRate < 0.10:
-		case lossRate < 0.30:
+		case r < 0.10:
+		case r < 0.30:
 			dev.MACState.DesiredParameters.ADRNbTrans = 2 + dev.MACState.DesiredParameters.ADRNbTrans/2
 		default:
```

Comment:
How about adding a filter for unrealistic values (and just use NbTrans=1 for those)? For example, we see ABP devices that increment a persistent FCnt by 1000 on every power cycle (because they don't want to persist the actual FCnt). We don't want to send all of those to NbTrans=3.